### ISSUE TITLE
Provider name optionality

### ISSUE THREAD

**Issue by ionut-arm** [parallaxsecond/parsec#571]
Posted as a comment by Matt in #570.

```
[[provider]]
name="mbed-crypto-provider"
```
I know we originally made the provider name globally optional so that existing configs would not fail with new versions of Parsec. Would it be better for it to be optional for `OnDisk` KIM implementations (to ensure config stability), and required for `SQLite` KIM configs? This way it forces the user to edit the config and set the provider names with their "preferred" naming scheme; preventing the (probably common) case of a user changing from the `OnDisk` to `SQLite` KIM without reading any of the provider naming warnings (quoted below).

In my mind, by going from `OnDisk` -> `SQLite` the user is "opting-in" to a new feature-set and its requirements. The only concern from the stability requirements set out is `New options should be optional`. I would argue that, transitively, provider name in this case is optional as the `manager_type="OnDisk"` would have to be mutated to `manager_type="SQLite"` in order for provider name to become a required field.

#### Parsec Config Stability:
> Old configuration files should still work with future stable Parsec versions, with the same default for optional options.
> 
> Configuration options should not disappear in future stable Parsec versions. Configuration defaults should remain the same. New options should be optional.

#### Current provider name warnings:

> ⚠ WARNING: Provider name cannot change.
> ⚠ WARNING: Choose a suitable naming scheme for your providers now.
> ⚠ WARNING: Provider name defaults to "mbed-crypto-provider" if not provided, you will not be able to change the provider's name from this if you decide to use the default.
> ⚠ WARNING: Changing provider name after use will lead to loss of existing keys.

On the note of these warnings. `WARNING: Provider name cannot change.` would probably read nicer as `WARNING: Provider name should not change once set.`. And maybe `loss of existing keys` to `LOSS OF EXISTING KEYS` to really drive the point home and draw the attention of skim readers.

_Originally posted by @MattDavis00 in https://github.com/parallaxsecond/parsec/pull/570#discussion_r780397295_

**Comment by ionut-arm** [comment]
>  a working deployment cannot be "switched" from on-disk to sql anyway

That's true, but it might be worth considering the "non-working-deployment" configs, like the default one in Fedora or openSUSE. Do we want those ones to be guided more stringently? Or is it better to collaborate with them on how the config file looks?

**Comment by paulhowardarm** [comment]
I would prefer to do this collaboratively - we need to be communicating the arrival of these new features so that distro maintainers can understand the impact on packaging - not just in terms of config tweaks, but also in terms of how to manage the new dependencies.

**Comment by ionut-arm** [comment]
Then I agree, doing this at the same time as #517 would make sense.

> It might be enough simply to ensure that our out-of-box example config includes provider names.

That's covered!

**Comment by MattDavis00** [comment]
My logic behind bringing this up applies to existing users, but was mainly focused on new users. With my main concern being users not choosing their provider naming scheme.

I'm in two minds about what I think:
- With new deployments, administrators should explicitly have to set the provider name (themselves) with their preferred naming scheme. E.g. leaving the provider name field blank for them to fill in or something similar. Mainly because we currently have no way (currently) to migrate keys from one `ProviderIdentity` to a new `ProviderIdentity`, if they did wish to rename their providers at a later date.
- Give provider names pre-filled as it is an easier installation experience, which is nice. This would ensure that all of the example configs can be run "as is". Main downside is users may not change their provider names when configuring the service, leaving them trapped with the default provider name that they did not explicitly choose, which may not match their naming scheme.

Overall though, my opinion is:

> That `provider.name` should not have an implicit default within the service, when using the  `SQLiteKeyInfoManager`; it must be set in `config.toml` visible to the user for transparency.
